Mh 253: handle cf responses

dsummersl opened the pull request:
What these changes do:
 * Added some `verification*` methods that can be used in the upload templates to display various text based on the results of the verification.
 * Added some simple styling based on Shane's comments ( I think a followon ticket will address some comps he will eventually give us?)
 * Implemented the callback handler endpoint that CF would call.
 * Opened up a couple options in the admin to make it easy to test out various states of a run.
 * I ran into some circular referencing with the 'cf run' model and opted to move it in with the other models that its closely related to.


---

- [x] Add test cases for views?
- [x] Add some example JSON docs that Gerald could use to test this. _I've updated the Jira ticket with notes for Gerald._
- [x] Provide some instructions for how the callback handler could be called (rather than the admin).
- [x] maybe I was too cavalier about moving the run model into documents...in hindsight seems like a big change. Maybe I should have hacked around the circular dependency instead?
- [x] When a document is uploaded it says 'saved' b/c the run hasn't been picked up immediately (and may not be). Perhaps something other than 'Saved' is in order akin to 'verifying' so it is clear what the next step is?
- [x] There is a risk that CF never calls our callback function...we should have a periodic task that updates the status of runs that have been 'processing' for 'too long'. Given that we don't have a working CF integration to truly test with, I think this should be a separate ticket (and if Shean that sounds right to you, I'll open that ticket on the backlog).
- [x] There is also no recovery from request failures or general failures - there should be some support for retries in some cases. I think this should also be a followup ticket, once we have a full verification process working with CF.

dsummersl committed:
[MH-253]: WIP Implements document handler view.

I think there are still some statuses that we want to store, or make
easily accessible to the views.

dsummersl committed:
[MH-253] Ensures data is stored from CF callback & show dashboard status.

Still something of a work in progress.

dsummersl committed:
[MH-253] Fixes test case for fn->property change.

dsummersl committed:
[MH-253] Refactors verified helper methods into AD model.

Also implements a number of factory methods to help with testing, and
refactors a couple tests to make use of them.

Also also moves the cloudfactory model into documents to avoid circular
references - these models are tightly bound as they are.

dsummersl committed:
[MH-253] Adds implementations for verification methods.

These methods will be used in the dashboard views to show the state of
verifiction.

dsummersl committed:
Merge branch 'myhpom-develop' into mh-253-handle-cf-responses

dsummersl committed:
[MH-253] Creates DocumentUrl and starts CF verification task on upload.

The actual template itself still needs styling and logic work.

dsummersl committed:
[MH-253] Ensure that the callback data has all the output we expect.

dsummersl committed:
[MH-253] Adds a 'failed' status for current status views.

This will be needed to notify the user when something bad has happened -
this is the worst case scenario where we'd hope for any of these
terminal states we have received an email...in the meantime the user is
asked to resubmit their document.

dsummersl committed:
[MH-253] Adds styling and language for specific errors.

dsummersl committed:
[MH-253] Fixes broken test.

dsummersl committed:
[MH-253] Adds language for detailed failed verifications.

dsummersl committed:
[MH-253] Turn off CSRF for the CF API endpoint.

sharrisoncaktus commented on myhpom/models/document.py:
```diff
@@ -75,8 +78,64 @@ def render_images(self, allpages=True, res=PDF_RESOLUTION, **gsargs):
+        # should be only one), return its output
+        processed_run = CloudFactoryDocumentRun.objects.filter(
+            status=CloudFactoryDocumentRun.STATUS_PROCESSED,
+            document_url__advancedirective=self).first()
```
I can foresee hitting unexpected errors, at which point we re-submit the AD to CF, creating a new run but not necessarily deleting the old. Should this .first() therefore be .last()?

dsummersl replied:
Good question - initially I'd implemented this as a 'look at all processed' runs for a document. But then I rethought it...shouldn't there only ever be one `STATUS_PROCESSED` or `STATUS_PROCESSING` run: if/when we reschedule a run b/c of a previous failure doesn't that mean the previous run(s) will be in a terminal state (DELETED, REQ_ERROR, etc)?

sharrisoncaktus replied:
Ah, yes, you're right – any other state will have been filtered out. I guess I'd still be more comfortable with .last(), but that's only because I don't trust the future.

dsummersl replied:
A scholarly thing to do - the future is never what we think! what about..not filtering and just doing 'get()' rather than first/last and changing the rest of these similar methods to use get() rather than exists()?

Then we'd have a fail-fast system - in the unlikely event that we have multiple of these records that should only be one - we get an exception, and an email?

sharrisoncaktus replied:
hmmm, I think that makes sense to me.

sharrisoncaktus commented on myhpom/models/document.py:
```diff
@@ -152,6 +211,220 @@ def authorized_client_ip(self, ip_address):
         return False
 
 
+class CloudFactoryDocumentRun(models.Model):
```
That's one way to solve the circular import problem.

dsummersl replied:
too brutal?

sharrisoncaktus replied:
Nah. Now that the Run class is closely associated with Document, it makes sense to put it here.

sharrisoncaktus commented on myhpom/templates/myhpom/upload/current_ad.html:
```diff
@@ -4,11 +4,53 @@
+
 {% block widget_body %}
 <dl class="advance-directive-widget__definition">
     <di>
```
I don't think `<di>` is an html tag? – you generally don't group the `<dh>` and `<dd>` together even though it seems like you should.

dsummersl replied:
ooh indeed...probably it should have been `<div>`?

dsummersl replied:
(looks like its allowed to group: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#Wrapping_name-value_groups_in_%3Cdiv%3E_elements)

sharrisoncaktus commented on myhpom/tests/factories.py:
```diff
@@ -25,3 +36,45 @@ class Meta(object):
     user_details = factory.RelatedFactory(UserDetailsFactory, 'user')
+
+
+class AdvanceDirectiveFactory(factory.django.DjangoModelFactory):
```
That had to happen, thanks for being the one!

sharrisoncaktus commented on myhpom/views/document.py:
```diff
@@ -41,7 +43,31 @@ def document_url(request, key):
+    except CloudFactoryDocumentRun.DoesNotExist:
+        raise Http404()
+    else:
+        # A run in a non-final state shouldn't be changed/updated again. Log an
```
change comment to "A run in a final state..."?

sharrisoncaktus commented on myhpom/templates/myhpom/upload/current_ad.html:
```diff
@@ -4,11 +4,53 @@
+        Could not be verified. Please re-scan and upload.
+    {% endif %}
+{% elif ad.verification_failed %}
+    Please re-scan and upload.
```
Did we agree (you, me, gannon on slack) that an error (probably ours) during the run process should not prompt the user to re-upload their document?

dsummersl replied:
No no agreement about any of this language. The ticket is pretty vague - I'll post what this looks like on the PR now and get Gannon's input.

sharrisoncaktus reviewed (changes requested):
I still have more review to do, but there are a few comments and questions to address.

dsummersl committed:
[MH-253] fixes tag typo.

dsummersl committed:
[MH-253] Fixes typo.

dsummersl committed:
[MH-253] Removes detail from 'failed' cases.

Here we don't want the user to resend their request until we have
directed them to do so.

dsummersl committed:
[MH-253] Updates status language.

dsummersl commented:
I posted the language to the ticket and updated relevant sections per Gannon's request.

sharrisoncaktus commented:
I was able to get a 500 error when reloading the dashboard after changing the status from "Processing" to "Processed" in the admin.

sharrisoncaktus commented:
I think moving the Run model in with the document model is fine. It *did* make it a little harder to review (every line is a green!), but since I'm very familiar with that model it wasn't that big a deal.

sharrisoncaktus commented:
I agree with changing "Saved" to something else – maybe show STATUS_NEW as "In Review"? I believe we've boxed that status as "the task is still in the queue."

sharrisoncaktus commented:
re/recovery and retry from request exceptions or other errors: I think we've done the best thing for now, which is to send us email and get us to review. As we see the different error states in testing with CF (or even later, in live work), we can add some retry logic if / when that seems appropriate.

sharrisoncaktus commented on myhpom/templates/myhpom/upload/current_ad.html:
```diff
@@ -4,17 +4,59 @@
+{% elif ad.verification_result %}
+    {% if ad.verification_result.owner_name_matches not in yes_or_na %}
+        Name on document doesn't match yours.
+    {% elif ad.verification_result.witness_signature_1 not in yes_or_na %}
```
I would just change each of these "elif" to "if" and show all the errors. Optionally do a `<ul>` in a smaller font

sharrisoncaktus commented on myhpom/views/document.py:
```diff
@@ -41,7 +43,31 @@ def document_url(request, key):
+        # A run in a final state shouldn't be changed/updated again. Log an
+        # error.
+        if run.status in CloudFactoryDocumentRun.STATUS_FINAL_STATES:
+            return HttpResponseBadRequest()
```
Include a reason / message in the response body? e.g., {"message": "Document run has already been processed and recorded."}

sharrisoncaktus commented:
For the case where CF accepts the run but never calls the callback, we could automate a timeout – 48 hours, use same timeout as for the URL – by posting an Abort and a new Submit for the document. This could be automated – scheduled task?

dsummersl committed:
[MH-253] Show all failed validations.

sharrisoncaktus commented on myhpom/views/document.py:
```diff
@@ -41,7 +43,31 @@ def document_url(request, key):
+        # try/catch here:
+        run.save_response_data(body)
+
+        return HttpResponse()
```
I would also include a message with this response, e.g, {"message": "Response data received, run completed."}

sharrisoncaktus commented on myhpom/templates/myhpom/upload/current_ad.html:
```diff
@@ -4,17 +4,65 @@
+        {% if ad.verification_result.signed_by_owner not in yes_or_na %}
+            <li>Not signed.</li>
+        {% endif %}
+        </ul>
```
Lookin' good!

dsummersl committed:
[MH-253] Ensure output() is resilient to unfamiliar JSON.

sharrisoncaktus commented on myhpom/models/document.py:
```diff
@@ -152,6 +211,220 @@ def authorized_client_ip(self, ip_address):
+            data = json.loads(self.response_content)
+            units = data['units']
+            return units[0]['output']
+        except ValueError:
```
If there is no 'output' block in the data, this raises a KeyError. Should we add 
```
        except KeyError:
            return None
```

sharrisoncaktus replied:
However, see below – maybe a better solution is to audit the callback request data to ensure the presence of an 'output' key.

sharrisoncaktus commented on myhpom/tests/test_cloudfactory_response_view.py:
```diff
@@ -0,0 +1,72 @@
+        # run to transition to finished.
+        failed_data = json.loads(SUCCESS_DATA)
+        for output in failed_data['units'][0]['output'].keys():
+            failed_data['units'][0]['output'][output] = False
```
I'm not confident that this test isolates _each_ key - the previous ones will still be false (python block scope). For example:
```
>>> d = {'a': True, 'b': True, 'c': True}
>>> for k in d:
...     d[k] = False; print d
... 
{'a': False, 'c': True, 'b': True}
{'a': False, 'c': False, 'b': True}
{'a': False, 'c': False, 'b': False}
>>> 
```

dsummersl replied:
Woops, I'll move the failed_data inside the loop.

dsummersl committed:
[MH-253] Refactors plain JSON post requests into common method.

sharrisoncaktus commented on myhpom/models/document.py:
```diff
@@ -152,6 +211,224 @@ def authorized_client_ip(self, ip_address):
+            has_required_keys = self.REQUIRED_OUTPUT_KEYS <= set(output.keys())
+            return all_true_or_na and has_required_keys
+        except ValueError:
+            return False
```
add a KeyError block here as well? 

With both of these, we have a situation where CloudFactory has returned a response without output. This seems to be an error condition that should raise our attention, set the run to STATUS_ERROR, and show the user the "In verification" message.

This suggests that we should audit the callback a little more so that we never get to the point of having a STATUS_PROCESSED that lacks an ['output'] key.

sharrisoncaktus commented on myhpom/templates/myhpom/upload/current_ad.html:
```diff
@@ -4,17 +4,65 @@
+{% if ad.verification_passed %}
+{% elif ad.verification_result %}
+    <div class="advance-directive-widget__verification-results">
+        <p class="advance-directive-widget__verification-results__title">Document is not valid:</p>
```
You could probably omit this `<p>` in light of the red "Invalid" right above it.

dsummersl replied:
ah yes, that is a tad redundant, I'll remove it.

dsummersl committed:
[MH-253] cleans up logic around CF response format.

dsummersl committed:
[MH-253] Makes save_response_data() more validate processed data.

This allows other methods to be a little simpler since they can assume
that the json that they are dealing with is the format we expect.

dsummersl committed:
[MH-253] Provide helpful HTTP error messages.

sharrisoncaktus commented:
So, here's a big question that I should have thought of earlier: Currently, if the admin deletes any completed runs, the user's document status changes from "Verified" to "Saved" – we lose the information about the verification of the document. Should we therefore save the document status in the AD itself rather than leaving it in the run? Or should we make it impossible for the admin to delete completed runs?

dsummersl committed:
[MH-253] Removes redundant title.

dsummersl committed:
[MH-253] Ensures that the test isolates each False output.

dsummersl committed:
[MH-253] Ensure admin-saved content doesn't break verification-* methods.

dsummersl commented:
Good question...Personally I think its good to remember how sharp the edges typically are in the django admin (and also personally, why I think its great for testers, not so great for actual users). Users can delete states now, and that too would probably be pretty bad if NC/SC were deleted. (We have yet to have any 'as an admin' inspired tickets as yet, probably since the CMS question will play a big part in this...)

We probably want to let testers continue to delete these runs, so that when we work on a follow up ticket to schedule cleanup states, they can simulate the kind of cleanup events we expect to happen (one of which would be an AD with no run at all)?

sharrisoncaktus reviewed (approved):
You fixed all the things and 500s, and good answer on the big question – admin for testers and us, not for Shane for example.
